### ISSUE TITLE
create_test concurrent tasks defaults on cori violate policies

### ISSUE THREAD

**Issue by brhillman** [E3SM-Project/E3SM#2923]
When a test suite is run on Cori, it seems that a large number of parallel builds are launched on the login nodes unless explicitly passing the `-jN` argument, where `N` is some reasonable number of concurrent tasks to launch. This will result in a ticket being opened and a nasty email from NERSC. It seems it would be better to have sane defaults that play nice with computing policies on the resources we use, maybe by setting defaults for the `-jN` argument.

Steps to reproduce:
```
./create_test e3sm_developer
```

Maybe this isn't a bug, and maybe this behavior is required for our nightly tests, but it's frustrating that running with defaults violates policies and results in an incident report.

**Comment by ndkeen** [comment]
Yes I agree and it's probably me who should be complaining/fixing.

Setting the default to -j3 will probably be fine for NERSC and us.

I actually think "something changed" in recent years that caused many more unix processes to be launched during our build than before. I started trying to capture some data using a clean compute haswell node, but did not finish.  We have 2 levels of parallelism -- the number of tests built at same time (-j sent to create_test) as well as the number of threads used within a component (GMAKE_J). However, the number of unix processes launched is FAR greater than expected.

**Comment by rljacob** [comment]
@jgfouca is the number of tests built at once a configurable parameter?

**Comment by jgfouca** [comment]
@rljacob only from the command line.

**Comment by jgfouca** [comment]
Oh, you can also configure it with .cime/config on a per-user basis.

**Comment by rljacob** [comment]
I think we need that in the machine config.

**Comment by jgfouca** [comment]
It wouldn't be hard to add a field for this; it's already using an XML value (MAX_MPITASKS_PER_NODE) as its default value.

**Comment by jgfouca** [comment]
Assigning @jhkennedy since this is a pretty easy upgrade to create_test.

**Comment by jhkennedy** [comment]
@brhillman I've got a PR up at ESMCI/cime#3115 to address this -- let me know if this doesn't work for you!

**Comment by brhillman** [comment]
@jhkennedy beautiful, thanks!

**Comment by jhkennedy** [comment]
This is *mostly* fixed now in ESMCI/CIME, but is likely going to be a recurrent issue as the number of parallel jobs are limited on a machine-by-machine basis due to constraints from CESM. To fix, set: 
```xml
<NTEST_PARALLEL_JOBS>4</NTEST_PARALLEL_JOBS>
```
for the affected machine in `cime/config/e3sm/machines/config_machines.xml`. 

Currently,  based on E3SM testing, I've set NTEST_PARALLEL_JOBS=4 on these E3SM machines:
* cori-haswell
* cori-knl
* blues
* anvil
* bebop
* theta
* titan
* summit

Other machines are still susceptible.

**Comment by rljacob** [comment]
Don't close the issue here on E3SM until the fix is merged to master.

**Comment by brhillman** [comment]
Re-opening this, because a `create_test e3sm_developer` with default args just got me a nasty email from NERSC staff for too many concurrent processes. I'm a little confused because even if I pass `-j 1` to `create_test`, I get at least 8 `fortcom` processes going. This is on cori-knl. Is there a different flag I need to pass to limit the parallel builds? I do not see anything in `./create_test --help`.

**Comment by brhillman** [comment]
Closing for now, my branch was out of date...

**Comment by jgfouca** [comment]
@brhillman , create_test supports multiple layers of parallelism. The first is how many cases will be processed simultaneously, this can be controlled via the -j argument that you are using. The second is the size of the processor core pool that create test uses, some phases use many cores and create_test will not pull in additional work if all the cores are being used, even if it's processing fewer than -j cases. The third is build parallelism within the case.build operation for a single case. In order to try to get reasonable build performance with the clunky (soon to be replaced) Makefile-based system, we currently launch several concurrent builds of components, each of which uses 'make -j$GMAKE_J'. This situation should be improved once we have CMake working, which does a much better job of exposing build parallelism.

**Comment by brhillman** [comment]
Okay, is there a better use for this on NERSC then, so that I can still build developer tests for my PRs and not keep getting reported?

**Comment by jgfouca** [comment]
@brhillman , with the current system, the only thing I can think of is to decrease GMAKE_J for cori in config_machines.xml. Once the CMake system is in place, the build system should not ever exceed GMAKE_J concurrent compilations for an invocation of case.build. If -j1 is passed, there should never be more than once case.build active.

Another thing to consider is just doing your builds on a compute node (if possible).

If I can get some guidance on how much CPU load NERSC would like us to stay below, that would help me configure things. It would also be nice if I could someday soon merge my CMake PR.

**Comment by rljacob** [comment]
Cori has been retired.
